src/collator.rs

```diff
@@ -70,24 +71,28 @@ pub(crate) enum SnapshotError {
     Reply(String),
 }
 
-pub(crate) struct Collator<F: crate::system_snapshot::SystemSnapshotter> {
-    system_snapshotter: F,
+pub(crate) struct Collator<F: crate::system_snapshot::SystemSnapshotter, P: crate::storage::Storage>
+{
+    system_snapshotter: Mutex<F>,
```

Comment:
1. It doesn't look like the `Mutex` here is ever held across an await point, so it's probably better to use the synchronous [`parking_lot::Mutex`](https://docs.rs/parking_lot/latest/parking_lot/type.Mutex.html) here, but see my next question:
2. Is there any reason snapshotters should be allowed to be internally unsynchronized? The default snapshot implementation doesn't do anything a mutex will help or hurt, making it seem odd to require that impl to pay the price of a mutex when it should be an implementation detail in extreme cases.